controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -194,18 +195,22 @@ func (ew *EndpointsWatcher) addService(obj interface{}) {
 }
 
 func (ew *EndpointsWatcher) deleteService(obj interface{}) {
-	service := obj.(*corev1.Service)
-	if service.Namespace == kubeSystem {
-		return
-	}
-	id := ServiceID{
-		Namespace: service.Namespace,
-		Name:      service.Name,
-	}
+	if serviceObj, err := extractDeletedObject(obj, reflect.TypeOf(&corev1.Service{})); err == nil {
+		service := serviceObj.(*corev1.Service)
+		if service.Namespace == kubeSystem {
+			return
+		}
+		id := ServiceID{
+			Namespace: service.Namespace,
+			Name:      service.Name,
+		}
 
-	sp, ok := ew.getServicePublisher(id)
-	if ok {
-		sp.deleteEndpoints()
+		sp, ok := ew.getServicePublisher(id)
+		if ok {
+			sp.deleteEndpoints()
+		}
+	} else {
+		ew.log.Error(err)
```

Comment:
Can we use the `utilruntime.HandleError()` to handle this, as seen [here](https://github.com/kubernetes/kubernetes/blob/0b2636a7e705bdf7c911e12d2e8846b9e6f88704/pkg/controller/replicaset/replica_set.go#L334-L335)? For more info of the function, see [here](https://godoc.org/k8s.io/apimachinery/pkg/util/runtime#HandleError).

Reply:
Correct me if I am wrong but the default behaviour of that is to just log. We are using a scoped logger here (withFields) I think and that would make it hard to diagnose where this problem actually occurs (if we are using utilruntime.HandleError()) instead? Or I guess we can add the information in the error message.

Reply:
[`utilruntime.HandleError()`](https://github.com/kubernetes/apimachinery/blob/fb45917aae8b47ab6f7c71d97863d287ee03bf6a/pkg/util/runtime/runtime.go#L82-L110) logs the error and perform some backoff. Now that I look closer at the source, it's using `klog`, which is different from our `ew.log`. Let's stick with what you have.